src/main/java/org/dependencytrack/tasks/repositories/ComposerMetaAnalyzer.java

```diff
@@ -73,76 +103,286 @@ public RepositoryType supportedRepositoryType() {
      */
     public MetaModel analyze(final Component component) {
         final MetaModel meta = new MetaModel(component);
+        final String composerPackageName = getComposerPackageName(component);
         if (component.getPurl() == null) {
             return meta;
         }
 
-        final String url = String.format(baseUrl + API_URL, urlEncode(component.getPurl().getNamespace()), urlEncode(component.getPurl().getName()));
+        final JSONObject repoRoot = getRepoRoot();
+        if (repoRoot == null) {
+            // absence of packages.json shouldn't happen, but let's try to get metadata as we did in <=4.12.2
+            return analyzeFromMetadataUrl(meta, component, PACKAGE_META_DATA_PATH_PATTERN_V1);
+        }
+
+        // Available packages is finite, so we can use it to determine if the package exists in this repository
+        if (repoRoot.has("available-packages")) {
+            final JSONArray availablePackages = repoRoot.getJSONArray("available-packages");
+            if (!availablePackages.toList().contains(getComposerPackageName(component))) {
+                return meta;
```

Comment:
I don't read it that way:
https://github.com/composer/composer/blob/fb397acaa0648ba2668893e4b786af6465a41696/doc/05-repositories.md?plain=1#L197

> If your repository only has a small number of packages, and you want to avoid
> the 404-requests, you can also specify an `"available-packages"` key in
> `packages.json` which should be **an array with all the package names that your **
> ** repository contains**. _Alternatively_ you can specify an
> `"available-package-patterns"` key which is an array of package name patterns
> (with `*` matching any string, e.g. `vendor/*` would make Composer look up
> every matching package name in this repository).

But the implementation seems to confirm your interpretation: 
https://github.com/composer/composer/blob/main/src/Composer/Repository/ComposerRepository.php#L1772

Reply:
Implemented